pkg/utils/addons/addons.go

```diff
@@ -0,0 +1,215 @@
+
+		taskDeploy, found, err := unstructured.NestedFieldNoCopy(newTempRes.Object, "spec", "tasks")
+		if !found || err != nil {
+			return pipelines, err
+		}
+		taskDeployMap := taskDeploy.([]interface{})[2].(map[string]interface{})
+		taskDeployMap["taskRef"] = map[string]interface{}{"name": "openshift-client", "kind": "ClusterTask"}
+		taskDeployMap["runAfter"] = []interface{}{"build"}
+		taskDeployMap["params"] = []interface{}{
+			map[string]interface{}{"name": "SCRIPT", "value": "kubectl $@"},
+			map[string]interface{}{"name": "ARGS", "value": []interface{}{"rollout", "status", "deploy/$(params.APP_NAME)"}},
+		}
+		pipelines = append(pipelines, newTempRes)
+	}
+	return pipelines, nil
+}
+
+func createKnativePipeline(template mf.Manifest, runtimes map[string]spec) ([]unstructured.Unstructured, error) {
+	environment := "knative"
+	var pipelines []unstructured.Unstructured
+	for name, spec := range runtimes {
+		labels := map[string]string{}
+		annotations := map[string]string{}
+		var params []interface{}
+		newTempRes := unstructured.Unstructured{}
+		template.Resources()[0].DeepCopyInto(&newTempRes)
+		labels[runtimeLabel] = spec.runtime
+		labels[typeLabel] = environment
+		annotations[preserveNSAnnotation] = "true"
+		if spec.supportedVersion != "" {
+			annotations[supportedVersionsAnnotation] = spec.supportedVersion
+		}
+		newTempRes.SetAnnotations(annotations)
+
+		newTempRes.SetName(name + "-knative")
+		newTempRes.SetLabels(labels)
+
+		tasks, found, err := unstructured.NestedFieldNoCopy(newTempRes.Object, "spec", "tasks")
+		if !found || err != nil {
+			return pipelines, err
+		}
+		taskBuild := tasks.([]interface{})[1].(map[string]interface{})
+		taskBuild["taskRef"] = map[string]interface{}{"name": name, "kind": "ClusterTask"}
+
+		params = append(params,map[string]interface{}{"name": "IMAGE", "value": "$(params.IMAGE_NAME)"} )
+		params = append(params,map[string]interface{}{"name": "TLSVERIFY", "value": "false"} )
+		if spec.version != "" {
+			params = append(params,map[string]interface{}{"name": "VERSION", "value": spec.version} )
+		}
+		if spec.minorVersion != "" {
+			params = append(params,map[string]interface{}{"name": "MINOR_VERSION", "value": spec.minorVersion} )
+		}
+		taskBuild["params"]=params
+
+		taskDeploy, found, err := unstructured.NestedFieldNoCopy(newTempRes.Object, "spec", "tasks")
+		if !found || err != nil {
+			return pipelines, err
+		}
+		taskDeployMap := taskDeploy.([]interface{})[2].(map[string]interface{})
+		taskDeployMap["name"] = "kn-service-create"
+		taskDeployMap["taskRef"] = map[string]interface{}{"name": "kn", "kind": "ClusterTask"}
+		taskDeployMap["runAfter"] = []interface{}{"build"}
+
+		taskDeployMap["params"] = []interface{}{
+			map[string]interface{}{"name": "ARGS", "value": []interface{}{"service", "create", "$(params.APP_NAME)", "--image=$(params.IMAGE_NAME)", "--force"}},
+		}
+		pipelines = append(pipelines, newTempRes)
+	}
+	return pipelines, nil
+}
```

Comment:
majority of the logic in these functions are the same. 
could you refactor these three functions, we can reuse common functionalities
![func-diff](https://user-images.githubusercontent.com/3650655/91026869-f909e400-e618-11ea-86f1-870c4375a78e.png)